homeassistant/components/zha/helpers.py

```diff
@@ -505,7 +507,14 @@ def __init__(
             DEBUG_LEVEL_CURRENT: async_capture_log_levels(),
         }
         self.debug_enabled: bool = False
-        self._log_relay_handler: LogRelayHandler = LogRelayHandler(hass, self)
+
+        log_relay_handler: LogRelayHandler = LogRelayHandler(hass, self)
+        log_simple_queue: queue.SimpleQueue[logging.Handler] = queue.SimpleQueue()
+        self._log_queue_handler = HomeAssistantQueueHandler(log_simple_queue)
+        self._log_queue_handler.listener = logging.handlers.QueueListener(
+            log_simple_queue, log_relay_handler
+        )
+
```

Comment:
After looking at this in detail, as far as I can tell there is no log file being written and everything ends up going out the websocket API on the other side of the dispatcher so since there is no potential for blocking I/O like the main log handler which is writing to `home-assistant.log`. Maybe I missed another log file being written?

Also it looks there doesn't seem to be a need to do the log formatting in another thread in the most expensive formatting is `entry.to_dict()` in `emit` process which is still likely cheaper that the cost to serialize it to json.

Reply:
If you could zip up the two SVG files posted above https://github.com/home-assistant/core/pull/138568#issuecomment-2660474010 it would be a lot easier to understand whats going on in the first one and if the highlighted purple the first svg are unexpected blocking operations or cpu intensive ones

Reply:
No, there is no disk i/o here. The intention was just to move the log formatting to another thread. I was under the impression that the motivation for the [original PR where the HomeAssistantQueueHandler](https://github.com/home-assistant/core/pull/35633) was also due to formatting? 

Added zip with SVGs to the comment above.

Reply:
I have not checked the zips yet, but I expect we don't have any expensive log formatting in ZHA since we control all the logging.  That problem is more likely to happen when don't know what's coming in and the underlying libraries or custom components have poorly optimized logging.

Reply:
> I have not checked the zips yet, but I expect we don't have any expensive log formatting in ZHA since we control all the logging.

Thanks for the zips. After looking, I agree that `__repr__ (zigpy/types/struct.py:392)` is a bit expensive so it makes sense to proceed with this.

An alternative, which would be even more performant, would be to speed up that function since we control that code and still do it in the event loop.  I don't know how viable that is though.

<img width="1328" alt="Screenshot 2025-02-17 at 9 36 05 AM" src="https://github.com/user-attachments/assets/fdb19a29-11f9-4ab2-93ab-ad8a6d08fcda" />
<img width="681" alt="Screenshot 2025-02-17 at 9 36 15 AM" src="https://github.com/user-attachments/assets/715a889e-d4c7-4da9-aec1-9f756f9e7b3a" />

Reply:
Here is the expensive code https://github.com/zigpy/zigpy/blob/4a3c058f8c15eb991c2c3cf0b03d137bcbd90456/zigpy/types/struct.py#L382